src/expr/type.cpp

```diff
@@ -326,7 +328,8 @@ bool Type::isTuple() const {
 /** Is this a record type? */
 bool Type::isRecord() const {
   NodeManagerScope nms(d_nodeManager);
-  return d_typeNode->isRecord();
+  return d_typeNode->getKind() == kind::DATATYPE_TYPE
+         && static_cast<DatatypeType>(*this).getDatatype().isRecord();
```

Comment:
Strictly speaking, this is undefined behavior because `*this` is not of type `DatatypeType` (UBSan should complain about this). The correct way is:

```suggestion
         && DatatypeType(*this).getDatatype().isRecord();
```